src/java_bytecode/java_bytecode_convert_method.cpp

```diff
@@ -505,6 +545,420 @@ void java_bytecode_convert_methodt::convert(
+    public graph<cfg_base_nodet<T, unsigned> >
+{
+  typedef std::map<unsigned, unsigned> entry_mapt;
+  entry_mapt entry_map;
+
+  procedure_local_cfg_baset() {}
+
+  void operator()(const address_mapt& amap)
+  {
+    for(const auto& inst : amap)
+    {
+      // Map instruction PCs onto node indices:
+      entry_map[inst.first]=this->add_node();
+      // Map back:
+      (*this)[entry_map[inst.first]].PC=inst.first;
+    }
+    for(const auto& inst : amap)
+    {
+      for(auto succ : inst.second.successors)
+        this->add_edge(entry_map.at(inst.first),entry_map.at(succ));
+    }
+  }
+
+  unsigned get_first_node(const address_mapt& amap) const { return amap.begin()->first; }
+  unsigned get_last_node(const address_mapt& amap) const { return (--amap.end())->first; }
+  unsigned nodes_empty(const address_mapt& amap) const { return amap.empty(); }
+
+};
+
+typedef java_bytecode_convert_methodt::local_variablet local_variablet;
+typedef std::map<local_variable_with_holest*,
+                 std::set<local_variable_with_holest*> > predecessor_mapt;
+
+struct is_predecessor_of
+{
+  const predecessor_mapt& order;
+
+  is_predecessor_of(const predecessor_mapt& _order) : order(_order) {}
+
+  bool operator()(local_variable_with_holest* a,
+                  local_variable_with_holest* b) const
+  {
+    auto findit=order.find(a);
+    if(findit==order.end())
+      return false;
+    return findit->second.count(b)>0;
+  }
+};
+
+static void gather_transitive_predecessors(
+  local_variable_with_holest* start,
+  const predecessor_mapt& predecessor_map,
+  std::set<local_variable_with_holest*>& result)
+{
+  if(!result.insert(start).second)
+    return;
+  auto findit=predecessor_map.find(start);
+  if(findit==predecessor_map.end())
+    return;
+  for(const auto pred : findit->second)
+    gather_transitive_predecessors(pred,predecessor_map,result);
+}
+
+static bool is_store_to_slot(
+  const java_bytecode_convert_methodt::instructiont& inst,
+  unsigned slotidx)
+{
+  const std::string prevstatement=id2string(inst.statement);
+  if(!(prevstatement.size()>=1 && prevstatement.substr(1,5)=="store"))
+    return false;
+
+  std::string storeslot;
+  if(inst.args.size()==1)
+  {
+    const auto& arg=inst.args[0];
+    storeslot=id2string(to_constant_expr(arg).get_value());
+  }
+  else
+  {
+    assert(prevstatement[6]=='_' && prevstatement.size()==8);
+    storeslot=prevstatement[7];
+    assert(isdigit(storeslot[0]));
+  }
+  auto storeslotidx=safe_string2unsigned(storeslot);
+  return storeslotidx==slotidx;
+}
+
+static void maybe_add_hole(
+  local_variable_with_holest& var,
+  unsigned from,
+  unsigned to)
+{
+  assert(to>=from);
+  if(to!=from)
+    var.holes.push_back({from,to-from});
+}
+
+void java_bytecode_convert_methodt::find_initialisers_for_slot(
+  local_variable_table_with_holest::iterator firstvar,
+  local_variable_table_with_holest::iterator varlimit,
+  const address_mapt& amap,
+  const java_cfg_dominatorst& doms)
+{
```

Comment:
This function is very long, would it be possible to factor out parts (e.g., loop bodies) into separate methods?